File: src/main/java/com/gtladd/gtladditions/integration/xei/LongFluidStack.java
package com.gtladd.gtladditions.integration.xei;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.level.material.Fluid;
import net.minecraftforge.fluids.FluidStack;

import com.google.common.primitives.Ints;
import org.jetbrains.annotations.NotNull;

public class LongFluidStack extends FluidStack {

    private long longAmount;

    public LongFluidStack(Fluid fluid, long amount) {
        super(fluid, Ints.saturatedCast(amount));
        this.longAmount = amount;
    }

    public LongFluidStack(Fluid fluid, long amount, CompoundTag nbt) {
        super(fluid, Ints.saturatedCast(amount), nbt);
        this.longAmount = amount;
    }

    public LongFluidStack(FluidStack stack, long amount) {
        super(stack, Ints.saturatedCast(amount));
        this.longAmount = amount;
    }

    public long getLongAmount() {
        return isEmpty() ? 0 : longAmount;
    }

    public void setLongAmount(long longAmount) {
        this.setAmount(Ints.saturatedCast(longAmount));
        this.longAmount = longAmount;
    }

    @Override
    public LongFluidStack copy() {
        return new LongFluidStack(getFluid(), longAmount, getTag());
    }

    @Override
    public boolean containsFluid(@NotNull FluidStack other) {
        return isFluidEqual(other) && other instanceof LongFluidStack longFluidStack ? this.longAmount >= longFluidStack.longAmount : this.longAmount >= other.getAmount();
    }

    @Override
    public boolean isFluidStackIdentical(FluidStack other) {
        return isFluidEqual(other) && other instanceof LongFluidStack longFluidStack ? this.longAmount == longFluidStack.longAmount : this.longAmount == other.getAmount();
    }
}

File: src/main/java/com/gtladd/gtladditions/integration/xei/LongFluidStack.kt
package com.gtladd.gtladditions.integration.xei

import com.google.common.primitives.Ints
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.level.material.Fluid
import net.minecraftforge.fluids.FluidStack

class LongFluidStack : FluidStack {
    private var longAmount: Long

    constructor(fluid: Fluid, amount: Long) : super(fluid, Ints.saturatedCast(amount)) {
        this.longAmount = amount
    }

    constructor(fluid: Fluid, amount: Long, nbt: CompoundTag?) : super(fluid, Ints.saturatedCast(amount), nbt) {
        this.longAmount = amount
    }

    constructor(stack: FluidStack, amount: Long) : super(stack, Ints.saturatedCast(amount)) {
        this.longAmount = amount
    }

    fun getLongAmount(): Long = if (isEmpty) 0 else longAmount

    fun setLongAmount(longAmount: Long) {
        this.amount = Ints.saturatedCast(longAmount)
        this.longAmount = longAmount
    }

    override fun copy(): LongFluidStack = LongFluidStack(fluid, longAmount, tag)

    override fun containsFluid(other: FluidStack): Boolean {
        return isFluidEqual(other) && if (other is LongFluidStack) {
            this.longAmount >= other.longAmount
        } else {
            this.longAmount >= other.amount
        }
    }

    override fun isFluidStackIdentical(other: FluidStack): Boolean {
        return isFluidEqual(other) && if (other is LongFluidStack) {
            this.longAmount == other.longAmount
        } else {
            this.longAmount == other.amount.toLong()
        }
    }
}